policy/rethinkdb/rethinkdb_test.go

```diff
@@ -0,0 +1,129 @@
+	"gopkg.in/ory-am/dockertest.v2"
+)
+
+var session *rdb.Session
+var s *Store
+
+var conditions = []policy.DefaultCondition{
+	{Operator: "foo", Extra: map[string]interface{}{"bar": "baz"}},
+	{Operator: "bar", Extra: map[string]interface{}{"foo": "baz"}},
+}
+
+var cases = []*policy.DefaultPolicy{
+	{uuid.New(), "description", []string{"user", "anonymous"}, policy.AllowAccess, []string{"article", "user"}, []string{"create", "update"}, conditions},
+	{uuid.New(), "description", []string{}, policy.AllowAccess, []string{"<article|user>"}, []string{"view"}, nil},
+	{uuid.New(), "description", []string{"<peter|max>"}, policy.DenyAccess, []string{"article", "user"}, []string{"view"}, conditions},
+	{uuid.New(), "description", []string{"<user|max|anonymous>", "peter"}, policy.DenyAccess, []string{".*"}, []string{"disable"}, conditions},
+	{uuid.New(), "description", []string{"<.*>"}, policy.AllowAccess, []string{"<article|user>"}, []string{"view"}, conditions},
+	{uuid.New(), "description", []string{"<us[er]+>"}, policy.AllowAccess, []string{"<article|user>"}, []string{"view"}, conditions},
+}
+
+func TestMain(m *testing.M) {
+	c, err := dockertest.ConnectToRethinkDB(20, time.Second, func(url string) bool {
+		rdbSession, err := rdb.Connect(rdb.ConnectOpts{
+			Address:  url,
+			Database: "hydra"})
+		if err != nil {
+			return false
+		}
+
+		_, err = rdb.DBCreate("hydra").RunWrite(rdbSession)
+		if err != nil {
+			return false
+		}
+
+		s = New(rdbSession)
+
+		if err := s.CreateTables(); err != nil {
+			return false
+		}
+
+		session = rdbSession
+
+		return true
+	})
+
+	if err != nil {
+		log.Fatalf("Could not connect to database: %s", err)
+	}
+
+	retCode := m.Run()
+
+	// force teardown
+	tearDown(session, c)
+
+	os.Exit(retCode)
+}
+
+func tearDown(session *rdb.Session, c dockertest.ContainerID) {
+	defer session.Close()
+	c.KillRemove()
+}
+
+func TestCreateErrors(t *testing.T) {
+	assert.NotNil(t, &policy.DefaultPolicy{ID: "invalid-format"})
+}
+
+func TestGetErrors(t *testing.T) {
+	_, err := s.Get(uuid.New())
+	assert.Equal(t, pkg.ErrNotFound, err)
+	_, err = s.Get("asdf")
+	assert.NotNil(t, err)
+}
+
+func TestCreateGetDelete(t *testing.T) {
+	for _, c := range cases {
+		err := s.Create(c)
+		assert.Nil(t, err)
+
+		get, err := s.Get(c.GetID())
+		assert.Nil(t, err)
+		pkg.AssertObjectKeysEqual(t, c, get, "Description", "Subjects", "Resources", "Effect", "Permissions")
+		assert.Equal(t, len(c.Conditions), len(get.GetConditions()))
+		//assert.True(t, reflect.DeepEqual(c, get), "%v does not equal %v", c, get)
+	}
+
+	for _, c := range cases {
+		assert.Nil(t, s.Delete(c.GetID()))
+		_, err := s.Get(c.GetID())
+		assert.NotNil(t, err)
+	}
+}
+
+func TestFindPoliciesForSubject(t *testing.T) {
+	for _, c := range cases {
+		require.Nil(t, s.Create(c))
+	}
+
+	policies, err := s.FindPoliciesForSubject("user")
+	assert.Nil(t, err)
+	assert.Equal(t, 5, len(policies))
+
+	policies, err = s.FindPoliciesForSubject("USER")
```

Comment:
Is this case-insensitive? Regex should (must?) be case sensitive. Because

* `[A-Z]+` matches `USER` but not `USER`
* `[a-z]+` matches `user` but not `USER`

This is important as some devs (YouTube!) use case sensitive IDs. So `UsEr` != `USER` != `user`.

I'm actually not sure if this is tested in postgres.